tethys_apps/base/controller.py

```diff
@@ -18,3 +48,286 @@ def as_controller(cls, **kwargs):
+        permissions_raise_exception: Raise 403 error if True. Defaults to False.
+
+    **Example:**
+
+    ::
+
+        from tethys_sdk.base import controller
+
+        @controller
+        def my_app_controller(request):
+            ...
+
+        @controller
+        def my_app_controller(request, url_arg):
+            ...
+
+        @controller(
+            name='custom_name',
+            url='customized-url/{url_arg}/with/arg',
+        )
+        def my_app_controller(request, url_arg):
+            ...
+
+        @controller
+        def my_app_controller(request, url_arg1, url_arg2=None, url_arg3=None):
+            ...
+
+        # Note: having arguments with default values in the controller function without specifying the ``url`` argument
+        # in the ``controller`` decorator will result in multiple ``UrlMap`` instances being created.
+        # In this case the following ``UrlMap`` instances would be generated:
+
+        [
+            UrlMap(
+                name='my_app_controller',
+                url='my-app-controller/{url_arg1}/'
+            ),
+            UrlMap(
+                name='my_app_controller_1',
+                url='my-app-controller/{url_arg1}/{url_arg2}/'
+            ),
+            UrlMap(
+                name='my_app_controller_2',
+                url='my-app-controller/{url_arg1}/{url_arg2}/{url_arg3}/'
+            )
+        ]
+
+        # Alternatively, you can explicitly define the names and urls generated by passing a dict as the ``url`` argument:
+
+        @controller(
+            url={
+                'custom_controller_name': 'custom-controller/{url_arg1/{url_arg2}/',
+                'another_custom_name': 'another-custom-controller/{url_arg1}/{url_arg3}/'
+            }
+        )
+        def my_app_controller(request, url_arg1, url_arg2=None, url_arg3=None):
+            ...
+
+        @controller(
+            name='controller_for_bokeh_handler',
+            handler='my_app.handlers.handler_func',
+            handler_type='bokeh',
+        )
+        def my_app_controller(request):
+            ...
+
+        @controller(
+            app_workspace=True,
+        )
+        def my_app_controller(request, app_workspace):
+            ...
+
+        @controller(
+            app_workspace=True,
+            user_workspace=True,
+        )
+        def my_app_controller(request, app_workspace, user_workspace, url_arg):
+            # Note that if both the ``app_workspace`` and ``user_workspace`` arguments are passed to the controller
+            # decorator, then "app_workspace" should preceed "user_workspace" in the function argument list,
+            # and both should be directly after the "request" argument.
+            ...
+
+        @controller(
+            login_required=False,
+        )
+        def my_app_controller(request):
+            # Note that ``login_required`` is ``True`` by default.
```

Comment:
"# Note that ``login_required`` is ``True`` by default (recommended). However, the login requirement is automatically dropped on all controllers when ``OPEN_PORTAL_MODE`` is enabled on the portal."